sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullFunctions.scala

```diff
@@ -68,22 +81,35 @@ case class Coalesce(children: Seq[Expression]) extends Expression {
     """ +
     children.map { e =>
       val eval = e.gen(ctx)
-      s"""
-        if (${ev.isNull}) {
-          ${eval.code}
-          if (!${eval.isNull}) {
-            ${ev.isNull} = false;
-            ${ev.primitive} = ${eval.primitive};
-          }
-        }
-      """
+      dataType match {
+        case DoubleType | FloatType =>
+          s"""
+            if (${ev.isNull}) {
+              ${eval.code}
+              if (!${eval.isNull} && !Double.isNaN(${eval.primitive})) {
```

Comment:
I think that `Double.isNaN` is implemented internally by taking a primitive `double` and checking whether `x != x`, so we might be able to just inline that check here rather than using the more clear static method.  However, do note that the `x != x` trick only work if you're dealing with non-boxed primitive doubles; `NaN = NaN` is true when you're dealing with boxed `Double` instances.